### ISSUE TITLE
Memory Leak in LineGeometry

### ISSUE THREAD

**Issue by giulioz** [mrdoob/three.js#24517]
**Describe the bug**

Calling the [`setPositions`](https://github.com/mrdoob/three.js/blob/9bb4782c22d7913ab49510b6ed8f51760f916759/examples/jsm/lines/LineSegmentsGeometry.js#L66) function of LineGeometry or LineSegmentsGeometry multiple times results in a memory leak that can crash the application. This happens because every time its called it creates a new Float32Array, InstancedInterleavedBuffer and two InterleavedBufferAttribute, instead of reusing the old ones.

**To Reproduce**

1. Create a scene with ~400 Line2s
2. During every frame, call `setPositions` on their geometry to update them
3. Let it run for some seconds
4. The entire browser freezes and the WebGL context gets lots

***Code***

```js
import * as THREE from "https://unpkg.com/three/build/three.module.js";
import { Line2 } from "https://unpkg.com/three/examples/jsm/lines/Line2.js";

var lines, renderer, scene, camera;

// Increase this number until it crashes (100 is enough for firefox)
const n = 5;

init();
animate();

function init() {
  renderer = new THREE.WebGLRenderer();
  renderer.setSize(window.innerWidth, window.innerHeight);
  renderer.setPixelRatio(window.devicePixelRatio);
  document.body.appendChild(renderer.domElement);

  scene = new THREE.Scene();

  camera = new THREE.PerspectiveCamera(
    40,
    window.innerWidth / window.innerHeight,
    0.1,
    10000
  );
  camera.position.set(0, 0, 4);

  lines = new Array(n).fill(0).map(() => new Line2());
  lines.forEach((line) => scene.add(line));
}

function animate(t) {
  requestAnimationFrame(animate);

  lines.forEach((line, i) => {
    line.material.resolution.set(window.innerWidth, window.innerHeight);
    const x = Math.cos(t + i / 20) * 2;
    const y = Math.sin(t + i / 20) * 2;
    line.geometry.setPositions([x, y, 0, x * 2 * (i / n), y * 2 * (i / n), 0]);
  });
  renderer.render(scene, camera);
}
```

***Live example***

* [jsfiddle-latest-release](https://jsfiddle.net/2hmrqtd6/13/)

**Expected behavior**

Not crash, reusing the same attributes.

**Screenshots**

Not necessary.

**Platform:**

 - Device: [Desktop]
 - OS: [MacOS]
 - Browser: [Chrome, Firefox]
 - Three.js version: [dev, r143]

**Comment by WestLangley** [comment]
`LineSegmentsGeometry.setPositions()`, and the other "set"  methods in the file, are intended to be called only once.

If you want to change the values, you have to update the buffer array values, not the buffer itself.

**Comment by WestLangley** [comment]
`LineSegmentsGeometry.setPositions()`, and the other "set"  methods in the file, are intended to be called only once.

If you want to change the values, you should update the buffer array values, not the buffer itself.

**Comment by giulioz** [comment]
I did that in my solution, but still it seems an abstraction leak having to rely on mutating the buffers directly, when a method could do that for you.

Also I couldn't find any info anywhere in the docs about the fact that "set" methods are intended to be called just once. What about calling it `setInitialPositions` or something like that, and adding a method `updatePositions` to do this without crashing the browser? I'm seeing a lot of code online that is calling `setPositions` every frame without knowing this detail, at least a comment could help.

For anyone interested, this is how I've done it:

```js
const attributeStarts = line.geometry.getAttribute("instanceStart");
for (let i = 0; i < points.length - 3; i += 3) {
  attributeStarts.array[2 * i] = points[i];
  attributeStarts.array[2 * i + 1] = points[i + 1];
  attributeStarts.array[2 * i + 2] = points[i + 2];

  attributeStarts.array[2 * i + 3] = points[i + 3];
  attributeStarts.array[2 * i + 4] = points[i + 4];
  attributeStarts.array[2 * i + 5] = points[i + 5];
}
attributeStarts.needsUpdate = true;
line.geometry.computeBoundingBox();
line.geometry.computeBoundingSphere();
```

**Comment by WestLangley** [comment]
>I couldn't find any info anywhere in the docs about the fact that "set" methods are intended to be called just once

`LineSegmentsGeometry.setPositions()`, and the other "set" methods in the file, are intended to be called only once.

This is an example, it is not part of the core library. You are free to modify your copy of the example.

> I'm seeing a lot of code online that is calling setPositions every frame

For this class?

**Comment by giulioz** [comment]
> `LineSegmentsGeometry.setPositions()`, and the other "set" methods in the file, are intended to be called only once.

> `LineSegmentsGeometry.setPositions()`, and the other "set" methods in the file, are intended to be called only once.

Well, I guess at least now there are at least two times on the web this detail is mentioned.
Closing the issue as there is now info on how to avoid this and a solution on the issue itself.

**Comment by WestLangley** [comment]
>I'm seeing a lot of code online that is calling setPositions every frame
>>For this class?

Can you please provide the links?

TBH, the OP is the first user I am aware of who has misused this method in this way. If there are others who are doing the same, it would be helpful to see the details. Doing so will help determine what change is warranted, if any.

**Comment by WestLangley** [comment]
>I'm seeing a lot of code online that is calling setPositions every frame
>>For this class?

Can you please provide the links?

TBH, the OP is the first user I am aware of who has used this method in this way. If there are others who are doing the same, it would be helpful to see the details. Doing so will help determine what change is warranted, if any.

**Comment by Mugen87** [comment]
Related #21488
